modules/godot/databags/databag_timer.cpp

```diff
@@ -0,0 +1,79 @@
+	ERR_FAIL_COND_V_MSG(p_end_in_seconds < 0.0, godex::TimerHandle(), "Can't make a new timer that ends at the past. To create an already timed out event, just enter 0.0 as time to end.");
+	return new_precise_timer(static_cast<uint64_t>(p_end_in_seconds * 1'000'000.0));
+}
+
+godex::TimerHandle TimersDatabag::new_precise_timer(const uint64_t p_end_in_microseconds) {
+	ERR_FAIL_COND_V_MSG(now + p_end_in_microseconds == 0, godex::TimerHandle(), "Can't make a new timer that ends at 0, it's invalid.");
+	if (destroyed_timers.size() > 0) {
+		godex::TimerHandle finalHandle = destroyed_timers[destroyed_timers.size() - 1];
+		internal_restart_precise_timer(finalHandle, now + p_end_in_microseconds);
+		destroyed_timers.remove_unordered(destroyed_timers.size() - 1);
+		return finalHandle;
+	} else {
+		timers.push_back(godex::Timer(now + p_end_in_microseconds, 0));
+		return godex::TimerHandle(timers.size() - 1, 0);
+	}
+}
+
+void TimersDatabag::restart_timer(const godex::TimerHandle p_timer_handle, const real_t p_end_in_seconds) {
+	ERR_FAIL_COND_MSG(p_end_in_seconds < 0.0, "Can't make a new timer that ends at the past. To create an already timed out event, just enter 0.0 as time to end.");
+	restart_precise_timer(p_timer_handle, static_cast<uint64_t>(p_end_in_seconds * 1'000'000.0));
+}
+
+void TimersDatabag::restart_precise_timer(const godex::TimerHandle p_timer_handle, const uint64_t p_end_in_microseconds) {
+	ERR_FAIL_COND_MSG(now + p_end_in_microseconds == 0, "Can't make a new timer that ends at 0, it's invalid.");
+	ERR_FAIL_COND_MSG(!is_valid_timer_handle(p_timer_handle), "The handle" + itos(p_timer_handle.timer_index) + " is not a valid handle.");
+	internal_restart_precise_timer(p_timer_handle, p_end_in_microseconds);
+}
+
+bool TimersDatabag::is_done(const godex::TimerHandle p_timer_handle) const {
+	ERR_FAIL_COND_V_MSG(!is_valid_timer_handle(p_timer_handle), true, "The handle" + itos(p_timer_handle.timer_index) + " is not a valid handle.");
+	return timers[p_timer_handle.timer_index].end_time <= now;
+}
+
+void TimersDatabag::destroy_timer(const godex::TimerHandle p_timer_handle) {
+	if (is_valid_timer_handle(p_timer_handle)) {
+		destroyed_timers.push_back(p_timer_handle);
+		destroyed_timers[destroyed_timers.size() - 1].generation++;
+		timers[p_timer_handle.timer_index].end_time = 0;
+		timers[p_timer_handle.timer_index].generation++;
+	}
+}
+
+real_t TimersDatabag::get_remaining_seconds(const godex::TimerHandle p_timer_handle) const {
+	ERR_FAIL_COND_V_MSG(!is_valid_timer_handle(p_timer_handle), 0, "The handle" + itos(p_timer_handle.timer_index) + " is not a valid handle.");
+	return static_cast<real_t>(static_cast<double>(timers[p_timer_handle.timer_index].end_time) - static_cast<double>(now) / 1'000'000.0);
+}
+
+uint64_t TimersDatabag::get_remaining_microseconds(const godex::TimerHandle p_timer_handle) const {
+	return is_done(p_timer_handle) ? 0 : timers[p_timer_handle.timer_index].end_time - now;
```

Comment:
I wonder if we should return `int64_t` and return the `-xx` when the timer expires. Still using the above suggested approach of calculating the remaining time. What do you think?

Reply:
No problem. Just thought that offering some interface capable of using the full timer precision would be nice, but I don't really think it would have many use-cases.

Reply:
The idea is to give the same API as `get_remaining_seconds` (that returns negatives) but with major precision. So the user don't need to use this any differently.